dtm0/recovery.c

```diff
@@ -0,0 +1,2170 @@
+		proc_conf = M0_CONF_CAST(obj, m0_conf_process);
+		rc = m0_conf_process2service_get(confc,
+						 &proc_conf->pc_obj.co_id,
+						 M0_CST_DTM0, &svc_fid);
+		if (rc != 0)
+			continue;
+
+		rc = recovery_fom_add(m, proc_conf, &svc_fid,
+				      is_svc_volatile(confc, &svc_fid));
+		if (rc != 0)
+			break;
+	}
+
+	m0_conf_diter_fini(&it);
+
+	/*
+	 * It is a workaround to propagate the current HA state.
+	 */
+	m0_confc_close(&root->rt_obj);
+	rc = m0_confc_root_open(confc, &root) ?:
+		m0_conf_diter_init(&it, confc,
+				   &root->rt_obj,
+				   M0_CONF_ROOT_NODES_FID,
+				   M0_CONF_NODE_PROCESSES_FID);
+	if (rc != 0)
+		goto out;
+
+	while ((rc = m0_conf_diter_next_sync(&it, conf_obj_is_process)) > 0) {
+		obj = m0_conf_diter_result(&it);
+		proc_conf = M0_CONF_CAST(obj, m0_conf_process);
+		rc = m0_conf_process2service_get(confc,
+						 &proc_conf->pc_obj.co_id,
+						 M0_CST_DTM0, &svc_fid);
+		if (rc != 0)
+			continue;
+
+		rf = recovery_fom_by_svc_find(m, &svc_fid);
+		if (rf == NULL)
+			break;
+
+		if (!rf->rf_is_local)
+			heq_post(rf, proc_conf->pc_obj.co_ha_state);
+	}
+
+	m0_conf_diter_fini(&it);
+	/* end of workaround */
+
+out:
+	if (root != NULL)
+		m0_confc_close(&root->rt_obj);
+	if (rc != 0)
+		unpopulate_foms(m);
+	return M0_RC(rc);
+}
+
+static struct m0_co_context *CO(struct m0_fom *fom)
+{
+	struct recovery_fom *rf = M0_AMB(rf, fom, rf_base);
+	return &rf->rf_coro;
+}
+
+#define F M0_CO_FRAME_DATA
+
+static void heq_await(struct m0_fom *fom, enum m0_ha_obj_state *out, bool *eoq)
+{
+	struct recovery_fom *rf = M0_AMB(rf, fom, rf_base);
+
+	M0_CO_REENTER(CO(fom),
+		      struct m0_be_op op;
+		      bool            got;
+		      uint64_t        state;);
+
+	F(got) = false;
+	F(state) = 0;
+	M0_SET0(&F(op));
+	m0_be_op_init(&F(op));
+	m0_be_queue_lock(&rf->rf_heq);
+	M0_BE_QUEUE_GET(&rf->rf_heq, &F(op), &F(state), &F(got));
+	m0_be_queue_unlock(&rf->rf_heq);
+	M0_CO_YIELD_RC(CO(fom), m0_be_op_tick_ret(&F(op), fom, RFS_WAITING));
+	m0_be_op_fini(&F(op));
+
+
+	if (F(got)) {
+		M0_LOG(M0_DEBUG, "heq_deq " FID_F " %s" ,
+		       FID_P(&rf->rf_tgt_svc),
+		       m0_ha_state2str(F(state)));
+		M0_ASSERT(ha_event_invariant(F(state)));
+		*out = F(state);
+	} else {
+		M0_LOG(M0_DEBUG, "heq_deq " FID_F, FID_P(&rf->rf_tgt_svc));
+		*eoq = true;
+	}
+}
+
+static void eolq_await(struct m0_fom *fom, struct eolq_item *out)
+{
+	struct recovery_fom *rf = M0_AMB(rf, fom, rf_base);
+
+	M0_CO_REENTER(CO(fom),
+		      struct m0_be_op op;
+		      bool            got;
+		      struct eolq_item item;);
+
+	F(got) = false;
+	M0_SET0(&F(item));
+	M0_SET0(&F(op));
+	m0_be_op_init(&F(op));
+	m0_be_queue_lock(&rf->rf_eolq);
+	M0_BE_QUEUE_GET(&rf->rf_eolq, &F(op), &F(item), &F(got));
+	m0_be_queue_unlock(&rf->rf_eolq);
+	M0_CO_YIELD_RC(CO(fom), m0_be_op_tick_ret(&F(op), fom, RFS_WAITING));
+	m0_be_op_fini(&F(op));
+
+	*out = F(got) ? F(item) : (struct eolq_item) { .ei_type = EIT_END };
+}
+
+static void restore(struct m0_fom *fom,
```

Comment:
The `heq_await()` and `eolq_await()` are straightforward to understand.
But this one is a little complex.
Can you add comments about this function?

Reply:
can we rename this function to 'recovery'?
Because this function is doing dtm recovery.

Reply:
Fixed/merged.